File: src/my_project/medium_problems/from51to100/next_permutation.py
from socket import NI_NUMERICHOST
from typing import List 


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        i = j = len(nums)-1
        while i > 0 and nums[i-1] >= nums[i]:
            i -= 1
        if i == 0:   # nums are in descending order
            nums.reverse()
            return 
        k = i - 1    # find the last "ascending" position
        while nums[j] <= nums[k]:
            j -= 1
        nums[k], nums[j] = nums[j], nums[k]  
        print(nums, 'sdfsdf')
        l, r = k+1, len(nums)-1  # reverse the second part
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            print(nums)
            l +=1 ; r -= 1
            
            
            
solution = Solution()

print(solution.nextPermutation([1,2,4,3]))
print(solution.nextPermutation([1,3,2,4]))

File: my_project_typescript/medium_problems_typescript/from1to50/nextPermutation.ts
function nextPermutation(nums: number[]): void {

    let i: number = nums.length -1 
    let j: number = nums.length -1 

    while (i > 0 && nums[i-1] >= nums[i]){
        i -= 1
    }
    if (i === 0){
        nums.reverse()
        return 
    }
    let k: number = i -1

    while (nums[j] <= nums[k]){
        j -= 1
    }
    let tempOne: number 
    let tempTwo: number 
    tempOne = nums[k]
    tempTwo = nums[j]
    nums[k] = tempTwo 
    nums[j] = tempOne

    let l: number = k+1
    let r: number = nums.length -1 

    while (l < r){
        tempOne = nums[l]
        tempTwo = nums[r]
        nums[l] = tempTwo
        nums[r] = tempOne
        l += 1
        r -= 1
    }
};